mobile/test/common/integration/client_integration_test.cc

```diff
@@ -869,6 +875,81 @@ TEST_P(ClientIntegrationTest, TimeoutOnResponsePath) {
   }
 }
 
+TEST_P(ClientIntegrationTest, ResetWithBidiTraffic) {
+  autonomous_upstream_ = false;
+  initialize();
+  ConditionalInitializer headers_callback;
+
+  stream_prototype_->setOnHeaders(
+      [this, &headers_callback](Platform::ResponseHeadersSharedPtr headers, bool,
+                                envoy_stream_intel) {
+        cc_.status = absl::StrCat(headers->httpStatus());
+        cc_.on_headers_calls++;
+        headers_callback.setReady();
+        return nullptr;
+      });
+
+  stream_->sendHeaders(envoyToMobileHeaders(default_request_headers_), false);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*BaseIntegrationTest::dispatcher_,
+                                                        upstream_connection_));
+  ASSERT_TRUE(
+      upstream_connection_->waitForNewStream(*BaseIntegrationTest::dispatcher_, upstream_request_));
+
+  // Send response headers but no body.
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  // Make sure the headers are sent up.
+  headers_callback.waitReady();
+  // Reset the stream.
+  upstream_request_->encodeResetStream();
+
+  // Encoding data should not be problematic.
+  Buffer::OwnedImpl request_data = Buffer::OwnedImpl("request body");
+  envoy_data c_data = Data::Utility::toBridgeData(request_data);
+  stream_->sendData(c_data);
```

Comment:
I'm a bit confused by this: once the stream is reset (in L904), I thought we shouldn't be allowed to send data on the stream?  Or is this the client->server stream whereas L904 cancels server->client?

Reply:
correct - once the stream is reset we shouldn't be allowed to send data on the stream.  
What can happen is
sendData is called but not executed
stream is reset, and reset is communicated to API
sendData is executed

from the mobile stack's point of view there's no way of differentiating between that and
stream is reset and reset is communicated to API
sendData is called
sendData is executed

the first set of ordering is almost impossible to guarantee so here we're intentionally breaking the API contract in order to test the race condition deterministcally.

Reply:
got it, thanks for the explanation!